pkg/exclusions/exclusions.go

```diff
@@ -0,0 +1,103 @@
+	"github.com/google/osv-scanner/pkg/lockfile"
+	"github.com/google/osv-scanner/pkg/reporter"
+)
+
+// ExcludePackages checks for matches in a Lockfile's Packages against the list of provided exclusion regex patterns.
+// If a match is found, the package is removed from the Lockfile's Packages.
+// The updated Lockfile is returned.
+func ExcludePackages(r reporter.Reporter, exclusionPatterns []string, parsedLockfile *lockfile.Lockfile) (*lockfile.Lockfile, error) {
+	// Create a new slice of packages to hold the filtered packages.
+	// Filtered packages are packages that do not match any of the exclusion patterns.
+	filteredPackages := make([]lockfile.PackageDetails, 0)
+
+	// Iterate through the packages in the Lockfile and check for matches against the exclusion patterns.
+	for _, pkg := range parsedLockfile.Packages {
+		matched, err := checkForRegexMatch(r, exclusionPatterns, pkg.Name)
+		if matched {
+			continue
+		}
+		if err != nil {
+			return parsedLockfile, err
+		}
+		filteredPackages = append(filteredPackages, pkg)
+	}
+	parsedLockfile.Packages = filteredPackages
+	// Return the updated Lockfile.
+	return parsedLockfile, nil
+}
+
+// ExcludeSBOMPackages checks if SBOM's ID matches the list of provided exclusion regex patterns.
+// If a match is found, the SBOM PURL is excluded from the OSV query by returning a boolean value of true to the calling scanSBOMFile function.
+func ExcludeSBOMPackages(r reporter.Reporter, exclusionPatterns []string, id *sbom.Identifier) (bool, error) {
+	matched, err := checkForRegexMatch(r, exclusionPatterns, id.PURL)
+	if err != nil {
+		return false, err
+	}
+	if matched {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// IsRegexPatterns checks if a list of regex patterns are valid regex patterns.
+func IsRegexPatterns(patterns []string) error {
+	for _, pattern := range patterns {
+		if !isValidRegex(pattern) {
+			return fmt.Errorf("%s", pattern)
+		}
+	}
+
+	return nil
+}
+
+// checkForRegexMatch checks if a package matches any of the exclusion regex patterns.
+func checkForRegexMatch(r reporter.Reporter, exclusionPatterns []string, pkg string) (bool, error) {
+	// Parse the exclusion patterns into a list of Regexp objects.
+	exclusionRegexes, err := parseExclusions(exclusionPatterns)
+	if err != nil {
+		r.PrintError(fmt.Sprintf("Error parsing exclusion patterns: %s\n", err))
+		return false, fmt.Errorf("%w", err)
+	}
+	// Iterate through the regex patterns and check for matches against the package identification.
+	for _, re := range exclusionRegexes {
+		matched := false
+		if re.MatchString(pkg) {
+			r.PrintText(fmt.Sprintf("Regex Match Found for exclusion pattern: '%s'\n", re.String()))
+			r.PrintText(fmt.Sprintf("Excluding package from OSV query: %s\n", pkg))
+			matched = true
+		}
+		if matched {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// isValidRegex checks if a regex pattern is valid.
+func isValidRegex(pattern string) bool {
+	_, err := regexp.Compile(pattern)
```

Comment:
Regexp compiling should be done in `cachedregexp` - currently that only has `MustCompile` but this is a reason for `Compile`, and is extra nice because we know its going to get `Compile`'d at least twice

Reply:
Hey mate, just going to play back my understanding to make sure I fully understand your suggestion.

Thanks for pointing out the `cachedregex` package; definitely makes sense to move the regex handling to that package. So for that reason:

- I will also create a `Compile` function for the `cachedregex` package which can be called by the `checkForRegexMatch` function:
```
// Compile takes a regex pattern and returns a list of Regexp objects.
func Compile(pattern string) (*regexp.Regexp, error) {

	re, err := regexp.Compile(pattern)
	if err != nil {
		return nil, fmt.Errorf("invalid regex pattern: %s", pattern)
	}
	return re, nil
}
```

-  I will move `exclusions.isValidRegex` to `cachedregex.IsValidRegex`, which will also make use of the new `Compile` function.

```
// IsValidRegex checks if a regex pattern is valid.
func IsValidRegex(pattern string) bool {
	_, err := Compile(pattern)
	return err == nil
}
```

I have also updated the `parseExclusions` function to utilise the `cachedregexp.MustCompile` fucntionality as that function doesn't care about errors?

```
// parseExclusions parses a list of exclusion patterns and returns a list of Regexp objects.
func parseExclusions(patterns []string) ([]*regexp.Regexp, error) {
	exclusionRegexes := make([]*regexp.Regexp, len(patterns))
	for i, pattern := range patterns {
		re := cachedregexp.MustCompile(pattern)
		exclusionRegexes[i] = re
	}

	return exclusionRegexes, nil
}
```

Am I roughly in the ballpark or way off?

Thanks for your time.

Reply:
Just create the `Compile` function in `cachedregexp` - `isValidRegex` should remain where it is

Reply:
Thanks @G-Rath, updated now. 👌